### DISCUSSION TITLE
Custom group component with specified space for children

### DISCUSSION THREAD

**Discussion by ghost** [xyflow/xyflow#2963] [Q&A] [answered]
Hello, as the title says I would like to know if there is a way to create a custom component in which I will have a dedicated space to show some info, and then be explicit on when to render the child nodes.

Here a simple concept of what I would like to do:

![Senza titolo-2023-01-09-1020](https://user-images.githubusercontent.com/74532875/228169118-eb232a66-d98f-4094-bccb-a8993cedc707.png)

Just to be clear, I would like for the child nodes to not be able to be dragged on that upper section!

**Comment by basile-savouret** [comment]
i am also struggling to create a custom group node. I could not find any code example on how to do it. Even in the source code the only code we are able to find is :
```TYPESCRIPT
const GroupNode = () => null;

GroupNode.displayName = 'GroupNode';

export default GroupNode;
```
in `\packages\core\src\components\Nodes\GroupNode.tsx`, the real code might be in the non open source package. 
That's why i think the documentation is missing a Custom group node page.

**Reply by bcakmakoglu** [reply]
There is no non-open-source package or code, where did you hear/see that? 😄 
That *is* the real code.

**Reply by basile-savouret** [reply]
It's a deduction, I assume that there is a non open source package based on the fact that there are paid features on the project that in my mind must be protected as the code examples are hidden in the documentation.

If i am wrong then i have no clue on how to make my custom group node work as the source code is only a component returning `null`. A specific page on the documentation would be awesome for those reasons !

**Reply by bcakmakoglu** [reply]
There are *no* paid features. A pro account only gives you access the certain *examples*, not to some secret paid package.
All the features are open source and in this repository.
The group node returns null because it doesn't really render anything besides a wrapping element. It's just used as a node with no handles whatsoever.

If you want your own Group node, just create a node like any other and style it in a way that it encapsulates other nodes.
Check out the [Sub Flow example](https://reactflow.dev/docs/examples/layout/sub-flows/) for that.

**Reply by basile-savouret** [reply]
Alright, thanks for clarifying this 😄 ! I have seen this example already but it is using the default group node not a custom one. I think the problem of @tmanuel94 and mine is that we want to achieve a custom group node with a [custom node](https://reactflow.dev/docs/examples/nodes/custom-node/). But how can i get the children and display them in my component ? How can i know how many are there in my group ?

**Reply by bcakmakoglu** [reply, marked as answer]
You can use something like this to find out what nodes are in the group / are children of a certain node.
This would be a simple "GroupNode" that just displays the labels of it's children.

```tsx
import { memo } from "react";
import { useStore } from "reactflow";

const GroupNode = ({ id }) => {
  const nodes = useStore((store) =>
    store.getNodes().filter((n) => n.parentNode === id)
  );

  return (
    <span>
      Children
      {nodes.map((n) => (
        <>
          <br />
          {n.data.label}
        </>
      ))}
    </span>
  );
};

export default memo(GroupNode);
```

[Here's a sandbox.](https://codesandbox.io/s/eloquent-payne-899054?file=/GroupNode.js:0-390)

**Reply by basile-savouret** [reply]
It makes much more sense now, i see how to do what i want 😄 . Thank you for your precious time. I wish you a great day !

**Reply by JiProchazka** [reply]
Hi,
I'm solving the same thing.

1. How do you render the (child)nodes of a custom type?
2. how do you prevent child nodes to be rendered twice (first in the < ReactFlow /> itself and second as the children of custom group node)?
